### PR TITLE
Integrate auth adapter for Facebook accountkit login

### PR DESCRIPTION
**Pull request by 6thfdwp**
First time PR to parse-server, try to learn and contribute back as our app also migrated to open source parse-server since last year. Hope to be part of this community to make it robust.

### CODE REVIEW

**Comment by 6thfdwp** [comment]
[commit#1e8112](https://github.com/parse-community/parse-server/pull/4434/commits/1e81124b691742b7552e331d1ccd3b03efbe459b) as required in Accountkit doc:   
>If you have enabled the Require App Secret setting in your app's dashboards, most calls that accept an account access token as a parameter will now require an additional **appsecret_proof** parameter to verify that these calls are coming from your own servers.
The app secret proof is a sha256 hash of your access token.

**Comment by montymxb** [comment]
@6thfdwp thanks for the contribution! We're always glad to have new contributors in the community 👍 . I'm rerunning CI now as it looks like your tests are good. I'll follow up when that's done.

File: src/Adapters/Auth/facebookaccountkit.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            reject(data.error);
+          } else {
+            resolve(data);
+          }
+        } catch (e) {
+          reject(e);
+        }
+      });
+    }).on('error', function () {
+      reject('Failed to validate this access token with Facebook.');
```
Comment (montymxb) [inline src/Adapters/Auth/facebookaccountkit.js]: You'll want to customize this error message as it's identical to the one for facebook auth.

File: src/Adapters/Auth/facebookaccountkit.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    return `me?access_token=${access_token}&appsecret_proof=${appsecret_proof}`
+  }
+  return `me?access_token=${access_token}`;
+}
+
+function validateAppId(appIds, authData, options) {
+  if (!appIds.length) {
+    throw new Parse.Error(
+      Parse.Error.OBJECT_NOT_FOUND,
+      'Facebook app id is not configured.');
```
Comment (montymxb) [inline src/Adapters/Auth/facebookaccountkit.js]: Same here

File: src/Adapters/Auth/facebookaccountkit.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  }
+  return graphRequest(getRequestPath(authData, options))
+    .then(data => {
+      if (data && data.application && appIds.indexOf(data.application.id) != -1) {
+        return;
+      }
+      throw new Parse.Error(
+        Parse.Error.OBJECT_NOT_FOUND,
+        'Facebook app id is invalid for this user.');
```
Comment (montymxb) [inline src/Adapters/Auth/facebookaccountkit.js]: and here

File: src/Adapters/Auth/facebookaccountkit.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+function validateAuthData(authData, options) {
+  return graphRequest(getRequestPath(authData, options))
+    .then(data => {
+      if (data && data.id == authData.id) {
+        return;
+      }
+      throw new Parse.Error(
+        Parse.Error.OBJECT_NOT_FOUND,
+        'Facebook auth is invalid for this user.');
```
Comment (montymxb) [inline src/Adapters/Auth/facebookaccountkit.js]: and also here

**Comment by 6thfdwp** [comment]
Thanks for the feedback! I'll do these tweaks.

I've also actually done some local test for full request flow, all good, but this requires real access token and secret which not suitable to put in test case. But surely I can add some faulty request case with bad token etc.

Have a nice holiday! 😀

File: src/Adapters/Auth/facebookaccountkit.js (hunk 2 above)
Comment (6thfdwp) [reply src/Adapters/Auth/facebookaccountkit.js]: Facebook app id is the same for Account Kit login, so I'll probably go like 'Facebook app id for Account Kit'

**Comment by 6thfdwp** [comment]
Hi @montymxb @flovilmart,

just check how to make this PR pass CI and other checks, I've done changes based on @montymxb review. I'd like to find new issues/enhancement to work on, hope that I can learn to get workflow right from the first PR for future commits.

Thanks!

**Comment by montymxb** [comment]
Hey @6thfdwp , been a bit absent 😆 , haven't been around for like 2 weeks...

I've reupped to the latest changes in master and we're pending some CI here now. Looks like the last run passed but I just want to verify that this one is good. If it succeeds I'll give this a final pre-flight inspection. 

Thanks for pinging us by the way! I track things _mostly_ asynchronously using the notifications we get  for the repos. Sometimes previous work can get buried while we're attempting to address what's new for a given week or day. I generally don't touch anything until I can actually respond to it, because as soon as I view anything the notification is cleared. Really wish github would allow us to reflag things for later followup or to make notifications unread.

File: spec/AuthenticationAdapters.spec.js [hunk 5]
```diff
@@ -345,4 +345,56 @@ describe('AuthenticationProviders', function() {
...
+    const {adapter} = authenticationLoader.loadAuthAdapter('facebookaccountkit', options);
+    adapter.validateAuthData(authData)
+      .then(done.fail, err => {
+        expect(err.code).toBe(190);
+        expect(err.type).toBe('OAuthException');
+        done();
+      })
+  });
+
+  it('should fail to validate Facebook accountkit auth with bad taken regardless of app secret proof', (done) => {
```
Comment (montymxb) [inline spec/AuthenticationAdapters.spec.js]: typo on `taken`, should be `token`

File: src/Adapters/Auth/facebookaccountkit.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    return `me?access_token=${access_token}&appsecret_proof=${appsecret_proof}`
+  }
+  return `me?access_token=${access_token}`;
+}
+
+function validateAppId(appIds, authData, options) {
+  if (!appIds.length) {
+    throw new Parse.Error(
+      Parse.Error.OBJECT_NOT_FOUND,
+      'Facebook app id for Account Kit is not configured.');
```
Comment (montymxb) [inline src/Adapters/Auth/facebookaccountkit.js]: I would add one more test to help cover this, just omit your appIds and this should trigger.

**Comment by montymxb** [comment]
I would like to add that other than those couple things the code in this PR looks good to me. Once those are addressed I believe we should be good to approve & merge.

**Comment by 6thfdwp** [comment]
@montymxb no worries, totally understand when facing lots of issues and requests.

I've actually used this adapter in our code base, it's been live supporting phone number login. I thought would be good to integrate and make more improvements.

Thanks for review again, I will push once I update.

File: src/Adapters/Auth/facebookaccountkit.js (hunk 6 above)
Comment (6thfdwp) [reply src/Adapters/Auth/facebookaccountkit.js]: Done.

**Comment by otymartin** [comment]
@6thfdwp is this live or dead?
Im looking to integrate accountkit

**Comment by 6thfdwp** [comment]
@otymartin seems still under review, waiting to be merged

**Comment by otymartin** [comment]
@flovilmart anyway we can get this approved (assuming it's production ready ) 🙏🙏

**Comment by flovilmart** [comment]
@6thfdwp thanks for the PR, can you please update the docs alongside the PR before we merge this?

https://github.com/parse-community/docs/blob/gh-pages/_includes/parse-server/third-party-auth.md

Thanks!

**Comment by 6thfdwp** [comment]
@flovilmart just updated docs. Please review. 

Thanks!

**Comment by flovilmart** [comment]
This is looking good! Thanks!

**Comment by otymartin** [comment]
Thanks @6thfdwp

**Comment by 6thfdwp** [comment]
Awesome, thanks @flovilmart 

Hope it will work, feedback welcomed @otymartin

**Comment by montymxb** [comment]
@6thfdwp thanks for hanging in there! Sorry I wasn't able to finish this up myself, but glad to see you got this in.

**Pull request merged**